### ISSUE TITLE
Azure App Service with Offline Sync: Max URL Length

### ISSUE THREAD

**Issue by OPunktSchmidt** [Azure/azure-mobile-apps-net-client#517]
We have a Xamarin.Forms App with uses Azure App Service with Offline Sync feature. While pulling new records from the server during the sync, multiple **HTTP - GET - Requests** are send. The URLs of the HTTP-GET-Requests are very long (beacause many parameters). Because that, we get "Bad Request" erros.

We tried to exceed the max url length, but this does not work. Does anyone have an idea how to send HTTP-GET-Requests with long URLs to Azure App Service?

Our solution so far (not working):

      <system.web>
    <httpRuntime targetFramework="4.5" maxRequestLength="524288" maxUrlLength="2097151" maxQueryStringLength="2097151" />
      </system.web>
      <system.webServer>
        <security>
          <requestFiltering>
        <requestLimits maxAllowedContentLength="536870912" maxUrl="2097151" maxQueryString="2097151"/>
          </requestFiltering>
        </security>
      </system.webServer>

**Comment by ltaljaard** [comment]
Is this related to when you write a linq query on the mobile app side (not the backend/server) that selects a range of items, and that then expands out into a very long SQL SELECT statement because each value in the range gets listed individually and "AND"-ed together, resulting in a very long query string which it then tries to send up to the server inside the Azure Offline Sync internals? Or is it something else?

**Comment by OPunktSchmidt** [comment]
Yeah, that's exactly what I mean. Look at the following code as an example.

If the documentPrimaryKeyIds list exceeds a certain number of entries, an HTTP Bad Request occurs.
I have to limit the number of list entries to prevent the error.

```
public async System.Threading.Tasks.Task DocumentListGetByPrimaryKeyIdsAsync(CPM.Azure.AppService.DataObjects.Client.Sync.List<CPM.Arda.Mobile.Freelancer.DataObjects.Entities.Document> list, System.DateTimeOffset? lastSyncDate, System.Threading.CancellationToken cancellationToken, System.Collections.Generic.List<System.Guid> documentPrimaryKeyIds)
{
    if (documentPrimaryKeyIds != null && documentPrimaryKeyIds.Count > 0)
    {
        Microsoft.WindowsAzure.MobileServices.IMobileServiceTableQuery<CPM.Arda.Mobile.Freelancer.DataObjects.Entities.Document> query =
            list.ISyncTable.Where(entity => (((entity.DeletionDate == null || entity.DeletionDate > System.DateTimeOffset.UtcNow) || (entity.DeletionDate != null && entity.DeletionDate > lastSyncDate && entity.Created < lastSyncDate))
                && (entity.NoSync == null || entity.NoSync > System.DateTimeOffset.UtcNow.Date))
                && documentPrimaryKeyIds.Contains(entity.PrimaryKeyId));

            await list.RemotePullAsync(query, null, cancellationToken).ConfigureAwait(false);
    }
}

```
This results in an HTTP-Get-Request like this:

> https://xxx:80/tables/Appointment?$filter=(((((DeletionDate%20eq%20null)%20or%20(DeletionDate%20gt%20datetimeoffset'2019-10-04T09%3A38%3A49.4556730%2B00%3A00'))%20or%20(((DeletionDate%20ne%20null)%20and%20(DeletionDate%20gt%20datetimeoffset'2019-10-04T08%3A23%3A23.5255190%2B00%3A00'))%20and%20(Created%20lt%20datetimeoffset'2019-10-04T08%3A23%3A23.5255190%2B00%3A00')))%20and%20((NoSync%20eq%20null)%20or%20(NoSync%20gt%20datetimeoffset'2019-10-04T00%3A00%3A00.0000000%2B02%3A00')))%20and%20(((((((((((((((PrimaryKeyId%20eq%20guid'7e1ac018-8219-4e72-bfb2-e2f732ff4ed8')%20or%20(PrimaryKeyId%20eq%20guid'66ee9ad8-b4b4-4b6a-bdff-3c7c9ac0b0b7'))%20or%20(PrimaryKeyId%20eq%20guid'd12d3884-a7c1-4093-a991-0322194e55b5'))%20or%20(PrimaryKeyId%20eq%20guid'3ebcb875-1e5b-4a10-b8c9-968ecbce8129'))%20or%20(PrimaryKeyId%20eq%20guid'98d9f640-d695-4a03-ad6e-78b91a02c37b'))%20or%20(PrimaryKeyId%20eq%20guid'879d2677-fedd-4920-81ab-71d79519fe68'))%20or%20(PrimaryKeyId%20eq%20guid'd363c02a-9779-4647-9e16-84e97e5af591'))%20or%20(PrimaryKeyId%20eq%20guid'06ccf440-1c59-41e6-b148-eb5725c7730b'))%20or%20(PrimaryKeyId%20eq%20guid'195302e9-536c-458b-9669-36e06e2eda9e'))%20or%20(PrimaryKeyId%20eq%20guid'810af477-f2e7-4e86-8d25-dbe5d6619f4d'))%20or%20(PrimaryKeyId%20eq%20guid'115d8a40-3ccd-4868-9707-bf9bed8f0e62'))%20or%20(PrimaryKeyId%20eq%20guid'4b456cc1-ad2c-4904-8884-ffdd29149a5a'))%20or%20(PrimaryKeyId%20eq%20guid'dabf6059-b996-423a-897c-37cc88e2542f'))%20or%20(PrimaryKeyId%20eq%20guid'e9ac63bc-f892-4a48-bfa1-25931f9022b1'))%20or%20(PrimaryKeyId%20eq%20guid'd18a27bd-56e3-4e44-ad73-ee1c8df09862')))&$skip=0&$top=1000&__includeDeleted=true

The more entries the list contains, the longer the url.

**Comment by ltaljaard** [comment]
Then you have my full sympathy, I have the same question. Would like to know how Microsoft suggests that this be handled 👍

**Comment by MatkovIvan** [comment]
Hi there,

The way with editing `web.config` to handle long URLs looks correct. I think it just isn't picked up by App Service for some reason, please check that this config is correct/parsable and that it is really applied.

As from client side, you can move `documentPrimaryKeyIds.Contains` filtering outside of actual query and do the check on client side (if it's acceptable of course)

Related docs:
- `maxQueryString` and `maxUrl` configuration: https://docs.microsoft.com/en-us/iis/configuration/system.webserver/security/requestfiltering/requestlimits/#configuration

Closing this issue as the problem (handling long URLs) isn't related to Azure Mobile Apps Xamarin/.NET Client SDK.
